<a href="https://colab.research.google.com/github/PauloBarretoDev/GenIaAtividade1/blob/main/AtividadeGenIA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tarefas para criar um RAG
1. Carregar os dados
2. Criar coluna para representar cada review
3. Criar coluna com embeddings
4. Criar função para buscar reviews relevantes a partir de uma query de busca
5. Criar prompt para perguntar ao gpt
6. Enviar prompt para o gpt e mostrar as respostas


## 1. Carregar os dados

In [ ]:
import pandas as pd;

df = pd.read_csv("/content/Games.csv");

## 2. Criar coluna para representar cada venda



In [ ]:
df["Name"] = df["Name"].astype(str);
df["Sales"] = df["Sales"].astype(float);
df["Publisher"] = df["Publisher"].astype(str);
df["texto_para_promtp"] = df["Name"] + " da publisher "+ df["Publisher"] + " vendeu "+ df["Sales"].astype(str) + " unidades na plataforma PC"


In [ ]:
!pip install python-dotenv


Collecting python-dotenv
  Downloading python_dotenv-1.0.0-py3-none-any.whl (19 kB)
Installing collected packages: python-dotenv
Successfully installed python-dotenv-1.0.0

In [ ]:
!pip install openai

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.0/77.0 kB 1.1 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
llmx 0.0.15a0 requires cohere, which is not installed.
llmx 0.0.15a0 requires tiktoken, which is not installed.


]
9s
!pip install openai
output
Collecting openai
  Downloading openai-0.28.1-py3-none-any.whl (76 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.0/77.0 kB 1.1 MB/s eta 0:00:00
Requirement already satisfied: requests>=2.20 in /usr/local/lib/python3.10/dist-packages (from openai) (2.31.0)
Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from openai) (4.66.1)
Requirement already satisfied: aiohttp in /usr/local/lib/python3.10/dist-packages (from openai) (3.8.6)
Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests>=2.20->openai) (3.3.1)
Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests>=2.20->openai) (3.4)
Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests>=2.20->openai) (2.0.7)
Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests>=2.20->openai) (2023.7.22)
Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp->openai) (23.1.0)
Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.10/dist-packages (from aiohttp->openai) (6.0.4)
Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /usr/local/lib/python3.10/dist-packages (from aiohttp->openai) (4.0.3)
Requirement already satisfied: yarl<2.0,>=1.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp->openai) (1.9.2)
Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from aiohttp->openai) (1.4.0)
Requirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.10/dist-packages (from aiohttp->openai) (1.3.1)
Installing collected packages: openai
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
llmx 0.0.15a0 requires cohere, which is not installed.
llmx 0.0.15a0 requires tiktoken, which is not installed.
Successfully installed openai-0.28.1

In [ ]:
import os
from dotenv import load_dotenv, find_dotenv

_ = load_dotenv(find_dotenv())

os.environ['OPENAI_API_KEY']
os.environ['OPENAI_API_BASE']

'https://openai-portodigital.openai.azure.com/'

https://openai-portodigital.openai.azure.com/

In [ ]:
import openai

openai.api_type = "azure"
openai.api_key  = os.environ['OPENAI_API_KEY']
openai.api_base = os.environ['OPENAI_API_BASE']
openai.api_version = "2023-05-15"

response = openai.Embedding.create(
    input="OI BRASIL",
    engine="text-similarity-ada-001"
)

In [ ]:

embeddings = response['data'][0]['embedding']
embeddings

[-0.019395481795072556,
 0.030103718861937523,
 -0.0001273331290576607,
 0.032396186143159866,
 0.02742917463183403,
 -0.0012279280927032232,
 0.02847486175596714,
 -0.013141470029950142,
 0.050112538039684296,
 -0.026041626930236816,
 0.00418526167050004,
 -0.026262830942869186,
 0.03752407431602478,
 -0.019174279645085335,
 -0.008516316302120686,
 -0.031229844316840172,
 -0.021336035802960396,
 -0.007430410478264093,
 -0.004117392469197512,
 -0.01598694548010826,
 0.043154697865247726,
 -0.013915680348873138,
 0.028434641659259796,
 -0.00904921442270279,
 -0.020421059802174568,
 0.018420178443193436,
 -0.031873345375061035,
 0.0047282143495976925,
 -0.01569535955786705,
 -0.01875198259949684,
 -0.04987122491002083,
 0.015946727246046066,
 0.0006541828042827547,
 0.06603915244340897,
 0.03975621610879898,
 -0.02026018500328064,
 -0.04027905687689781,
 -0.011854470707476139,
 0.011502556502819061,
 0.01390562579035759,
 -0.045608039945364,
 -0.00938604585826397,
 0.001997111365199089,


output
[-0.019395481795072556,
 0.030103718861937523,
 -0.0001273331290576607,
 0.032396186143159866,
 0.02742917463183403,
 -0.0012279280927032232,
 0.02847486175596714,
 -0.013141470029950142,
 0.050112538039684296,
 -0.026041626930236816,
 0.00418526167050004,
 -0.026262830942869186,
 0.03752407431602478,
 -0.019174279645085335,
 -0.008516316302120686,
 -0.031229844316840172,
 -0.021336035802960396,
 -0.007430410478264093,
 -0.004117392469197512,
 -0.01598694548010826,
 0.043154697865247726,
 -0.013915680348873138,
 0.028434641659259796,
 -0.00904921442270279,
 -0.020421059802174568,
 0.018420178443193436,
 -0.031873345375061035,
 0.0047282143495976925,
 -0.01569535955786705,
 -0.01875198259949684,
 -0.04987122491002083,
 0.015946727246046066,
 0.0006541828042827547,
 0.06603915244340897,
 0.03975621610879898,
 -0.02026018500328064,
 -0.04027905687689781,
 -0.011854470707476139,
 0.011502556502819061,
 0.01390562579035759,
 -0.045608039945364,
 -0.00938604585826397,
 0.001997111365199089,
 0.008687245659530163,
 0.0128901032730937,
 -0.022804019972682,
 -0.027911799028515816,
 -0.04395907372236252,
 -0.007309754379093647,
 -0.015504320152103901,
 -0.01229687687009573,
 0.028133001178503036,
 -0.005866907071322203,
 0.002352795796468854,
 0.0010104955872520804,
 -0.023346973583102226,
 0.022884458303451538,
 -0.05373222380876541,
 -0.015554593876004219,
 0.022844238206744194,
 -0.008134238421916962,
 -0.0022132869344204664,
 0.019717233255505562,
 -0.018651437014341354,
 -0.03497018665075302,
 0.02899770438671112,
 -0.0005467233713716269,
 0.0011814251774922013,
 -0.029399892315268517,
 0.0012932835379615426,
 0.026423705741763115,
 0.01016025710850954,
 -0.01613776572048664,
 0.004584935028105974,
 0.002003395464271307,
 0.0027323600370436907,
 -0.02085341140627861,
 -0.04552760347723961,
 0.02258281782269478,
 0.007535984739661217,
 0.0007440465269610286,
 0.011924853548407555,
 0.02017974853515625,
 -0.007465601898729801,
 0.02789168991148472,
 0.04850379005074501,
 -0.02159745804965496,
 -0.025760097429156303,
 -0.021134942770004272,
 -0.025840533897280693,
 0.04568847641348839,
 0.0022308826446533203,
 -0.015092078596353531,
 -0.035392481833696365,
 -0.01801799051463604,
 -0.057955190539360046,
 0.00967763178050518,
 -0.013986063189804554,
 -0.011050095781683922,
 0.01539371907711029,
 0.01659022644162178,
 0.011924853548407555,
 -0.010969658382236958,
 -0.02085341140627861,
 0.016841594129800797,
 -0.013865407556295395,
 0.006510406732559204,
 -0.005293790251016617,
 -0.018490560352802277,
 0.00023801319184713066,
 0.04621132090687752,
 0.04432104155421257,
 -0.03619685769081116,
 -0.02497583068907261,
 0.023869816213846207,
 0.03350220248103142,
 -0.003137060906738043,
 0.02563944086432457,
 -0.013855352997779846,
 0.026725346222519875,
 -0.016258422285318375,
 -0.05188216269016266,
 0.07452530413866043,
 0.013845297507941723,
 0.0070382775738835335,
 0.014941258355975151,
 0.006414887495338917,
 -0.0017331759445369244,
 -0.006751719396561384,
 -0.022623036056756973,
 0.002966131316497922,
 0.020209912210702896,
 -0.014217320829629898,
 -0.008998940698802471,
 -0.013815133832395077,
 0.002338970545679331,
 -0.026363378390669823,
 0.05570293962955475,
 0.043757978826761246,
 -0.035392481833696365,
 0.04580913484096527,
 -0.0019015917787328362,
 0.0026896276976913214,
 -0.028153110295534134,
 -0.010125065222382545,
 0.013885516673326492,
 -0.03623707592487335,
 -0.012759392149746418,
 0.034266356378793716,
 -0.02946021966636181,
 0.020591989159584045,
 -0.01125118974596262,
 -0.0366191528737545,
 0.061936844140291214,
 0.01151261106133461,
 -0.0019405537750571966,
 -0.00489663053303957,
 0.020612100139260292,
 0.0040771737694740295,
 0.0015622463542968035,
 -0.016318749636411667,
 -0.016047272831201553,
 0.0035568438470363617,
 0.00997927226126194,
 0.025016050785779953,
 -0.008817956782877445,
 -0.025760097429156303,
 0.01524289883673191,
 -0.009647468104958534,
 0.011623213067650795,
 -0.04822225868701935,
 0.0001861687342170626,
 -0.011643322184681892,
 -0.0010450585978105664,
 -0.005308872554451227,
 -0.021577348932623863,
 -0.004939362872391939,
 0.01196507178246975,
 -0.051158223301172256,
 0.0071639614179730415,
 0.029037922620773315,
 -0.02519703470170498,
 0.03149126470088959,
 0.013583876192569733,
 0.015886398032307625,
 0.009019049815833569,
 0.023005113005638123,
 -0.01363414991647005,
 0.03271793574094772,
 -0.0014290217077359557,
 0.01988816261291504,
 0.0007000572513788939,
 0.06314340978860855,
 0.046533070504665375,
 -0.012055564671754837,
 -0.03155159205198288,
 0.042350322008132935,
 -0.028233548626303673,
 -0.014227375388145447,
 -0.008083964698016644,
 0.004107337910681963,
 0.0070382775738835335,
 0.03257717192173004,
 0.004021873231977224,
 0.007259480655193329,
 0.011593049392104149,
 0.0073700821958482265,
 -0.011904744431376457,
 0.01727394387125969,
 0.007314781658351421,
 0.042189449071884155,
 -0.06097159534692764,
 0.00045214651618152857,
 0.018329685553908348,
 -0.02071264572441578,
 -0.0030918149277567863,
 -0.011904744431376457,
 -0.04677438363432884,
 0.022080082446336746,
 0.05554206669330597,
 0.006842211354523897,
 -0.014850766398012638,
 -0.032396186143159866,
 -0.01585623435676098,
 0.03806702792644501,
 -0.029198797419667244,
 -0.037121888250112534,
 0.037121888250112534,
 -0.0021894071251153946,
 0.02527747116982937,
 0.03332121670246124,
 -0.0003607117396313697,
 0.028977595269680023,
 -0.005982535891234875,
 -0.006158492993563414,
 -0.0024319763761013746,
 0.04460257291793823,
 0.0102809127420187,
 0.026504144072532654,
 0.0321347638964653,
 0.013262126594781876,
 -0.051158223301172256,
 -0.005449637770652771,
 0.04609066620469093,
 -0.010999822989106178,
 -0.001543393824249506,
 -0.015645086765289307,
 -0.02017974853515625,
 0.028816720470786095,
 0.03368318825960159,
 -0.03131028264760971,
 -0.014036336913704872,
 0.04818204045295715,
 0.0010255776578560472,
 0.008104073815047741,
 0.01703263260424137,
 -0.026182392612099648,
 -0.0023138339165598154,
 0.03581478074193001,
 0.04460257291793823,
 0.015866288915276527,
 0.01972728781402111,
 -0.027228079736232758,
 -0.0008628174546174705,
 -0.021858880296349525,
 0.007425383199006319,
 -0.014961367473006248,
 0.03649849817156792,
 0.004449197091162205,
 0.0017369464039802551,
 -0.034125592559576035,
 0.001589896623045206,
 -0.013915680348873138,
 -0.008400687016546726,
 0.020521607249975204,
 0.015162461437284946,
 -0.005665813572704792,
 -0.029118360951542854,
 0.03649849817156792,
 0.05433550477027893,
 0.015423882752656937,
 0.019405538216233253,
 -0.023507848381996155,
 0.010029545985162258,
 -0.008275004103779793,
 0.008083964698016644,
 -0.009662549942731857,
 -0.03221520408987999,
 0.029198797419667244,
 -0.005856852512806654,
 0.04066113755106926,
 0.013513493351638317,
 0.04303404316306114,
 -0.2916662395000458,
 -0.009003967978060246,
 -0.05711059644818306,
 0.04677438363432884,
 0.028494970872998238,
 0.017957663163542747,
 -0.003984168171882629,
 -0.00982845202088356,
 0.010637854225933552,
 0.02869606390595436,
 -0.006399805657565594,
 0.004657831508666277,
 0.011331627145409584,
 0.005464720074087381,
 0.021557239815592766,
 0.007535984739661217,
 0.005846797954291105,
 -0.004479361232370138,
 -0.002219571266323328,
 -0.0022208280861377716,
 0.0028203383553773165,
 -0.014659726992249489,
 -0.020370787009596825,
 -0.009139706380665302,
 -0.015564648434519768,
 -0.03301957622170448,
 0.059322625398635864,
 0.005675868131220341,
 -0.03273804485797882,
 0.03484952822327614,
 0.04210900887846947,
 0.008360468782484531,
 0.008028663694858551,
 0.030586345121264458,
 -0.01964684948325157,
 0.01637907698750496,
 0.05518009886145592,
 -0.025619331747293472,
 -0.0014001145027577877,
 -0.022743692621588707,
 0.005409419070929289,
 -0.029781969264149666,
 0.04355688393115997,
 0.00026754880673252046,
 0.01622825674712658,
 0.0015534484991803765,
 -0.010607690550386906,
 -0.025599222630262375,
 -0.006600899156183004,
 0.022482270374894142,
 -0.0075460392981767654,
 -0.048624444752931595,
 -0.011221026070415974,
 0.018329685553908348,
 -0.07826565206050873,
 0.056386660784482956,
 0.02258281782269478,
 -0.0017319191247224808,
 0.03416581079363823,
 -0.03488975018262863,
 0.051681067794561386,
 0.02638348750770092,
 0.002367877634242177,
 -0.0001880539784906432,
 0.0029435083270072937,
 -0.049308162182569504,
 -0.0028178247157484293,
 0.01837996020913124,
 -0.005494884215295315,
 0.014589344151318073,
 -0.01913406141102314,
 0.015172515995800495,
 0.08494196087121964,
 0.0016225744038820267,
 0.01065796334296465,
 -0.03378373384475708,
 0.022240957245230675,
 -0.03824801370501518,
 0.01777667924761772,
 -0.03893173113465309,
 -0.012387368828058243,
 -0.028675954788923264,
 0.013231961987912655,
 0.011633267626166344,
 0.04898641258478165,
 0.02203986421227455,
 0.011432173661887646,
 0.07677755504846573,
 -0.0028228522278368473,
 -0.030103718861937523,
 -0.01083894819021225,
 0.011864525265991688,
 -0.013121360912919044,
 0.0022233417257666588,
 -0.03519139066338539,
 0.0183196309953928,
 0.02742917463183403,
 0.013402892276644707,
 0.05578337982296944,
 0.0033431819174438715,
 -0.0321347638964653,
 0.032999467104673386,
 0.0027122506871819496,
 -0.0388512946665287,
 0.014629563316702843,
 0.00975304190069437,
 0.010849002748727798,
 -0.008420797064900398,
 0.004946903791278601,
 -0.04170682281255722,
 0.007003086619079113,
 -0.03907249495387077,
 -0.0017671104287728667,
 -0.04275251179933548,
 0.02017974853515625,
 0.02332686260342598,
 -0.013131415471434593,
 0.014810547232627869,
 -0.009521784260869026,
 -0.001170742092654109,
 -0.04146550968289375,
 -0.005132915452122688,
 -0.010426705703139305,
 0.003974113147705793,
 0.023749159649014473,
 -0.0388714037835598,
 -0.0463319756090641,
 0.03832845017313957,
 0.01521273422986269,
 0.025538893416523933,
 0.04403951019048691,
 0.050273410975933075,
 -0.017907390370965004,
 -0.016409242525696754,
 0.016479624435305595,
 -0.025297582149505615,
 -0.02308555133640766,
 -0.03191356360912323,
 -0.009139706380665302,
 0.01110036950558424,
 0.019707178696990013,
 -0.024835065007209778,
 -0.016107602044939995,
 -0.004858925472944975,
 -0.02473451942205429,
 0.03161192312836647,
 0.033823952078819275,
 -0.0013485843082889915,
 -0.00900899525731802,
 -0.0033029632177203894,
 0.005394337233155966,
 0.04198835417628288,
 0.015936672687530518,
 -0.03179290518164635,
 0.028515079990029335,
 0.04116386920213699,
 0.01733427308499813,
 0.016107602044939995,
 -0.0038584843277931213,
 -0.005570294335484505,
 0.02011941932141781,
 -0.029661312699317932,
 -0.04983100667595863,
 -0.037041451781988144,
 0.006721555255353451,
 0.010190420784056187,
 0.003084273776039481,
 -0.024251895025372505,
 0.00967763178050518,
 -0.010849002748727798,
 -0.03746374696493149,
 -0.05099735036492348,
 -0.025579111650586128,
 0.012266712263226509,
 0.012196330353617668,
 0.028133001178503036,
 0.01225665770471096,
 -0.0224219411611557,
 0.007882870733737946,
 0.041666604578495026,
 0.008033690974116325,
 0.03368318825960159,
 -0.024312222376465797,
 -0.007425383199006319,
 0.01718345284461975,
 0.009969217702746391,
 0.0314108282327652,
 -0.030988531187176704,
 0.02489539422094822,
 -0.016811428591609,
 0.02340730093419552,
 0.01973734237253666,
 -0.06535543501377106,
 0.015494265593588352,
 -0.0212958175688982,
 -0.01199523638933897,
 0.06000634655356407,
 0.04552760347723961,
 -0.02630304917693138,
 -0.013654259033501148,
 0.04291338473558426,
 0.02803245559334755,
 0.0071639614179730415,
 0.04202857241034508,
 0.00025450915563851595,
 -0.02885693870484829,
 -0.052887629717588425,
 -0.02437254972755909,
 -0.06515434384346008,
 0.0038660254795104265,
 0.0717099979519844,
 -0.011412064544856548,
 0.01435808651149273,
 -0.01816881075501442,
 -0.046533070504665375,
 0.0013448138488456607,
 0.0017294053686782718,
 0.07862761616706848,
 0.004788542632013559,
 -0.02009931020438671,
 0.010758510790765285,
 0.006515434477478266,
 -0.0037076640874147415,
 0.018550889566540718,
 0.048624444752931595,
 0.005444610491394997,
 0.022804019972682,
 -0.004285808652639389,
 -0.004288322292268276,
 -0.023970363661646843,
 0.003971599508076906,
 -0.016399187967181206,
 0.039715997874736786,
 -0.0015283117536455393,
 -0.010200475342571735,
 0.013543657027184963,
 -0.021275708451867104,
 0.01793755404651165,
 0.009778179228305817,
 -0.04275251179933548,
 -0.011633267626166344,
 0.005952372215688229,
 -0.06636090576648712,
 -0.022844238206744194,
 0.03342176601290703,
 -0.016107602044939995,
 -0.04777985066175461,
 0.03951490297913551,
 -0.0029208853375166655,
 -0.02115505188703537,
 0.017676131799817085,
 0.058719344437122345,
 -0.05047450587153435,
 -0.016479624435305595,
 0.014740164391696453,
 -0.017746515572071075,
 0.006103192456066608,
 0.03116951510310173,
 -0.0030365141574293375,
 0.0075460392981767654,
 0.03754418343305588,
 -0.026886221021413803,
 0.006113247014582157,
 0.01785711571574211,
 0.018932966515421867,
 0.006304285954684019,
 0.029540657997131348,
 -0.008300140500068665,
 -0.02338719181716442,
 -0.03989697992801666,
 -0.05107778683304787,
 0.009898834861814976,
 -0.015936672687530518,
 0.04005785658955574,
 -0.035995762795209885,
 -0.14816580712795258,
 0.035995762795209885,
 -0.01696224883198738,
 0.006314340513199568,
 -0.04822225868701935,
 0.012869993224740028,
 0.021718114614486694,
 -0.033904388546943665,
 0.03529193624854088,
 -0.024935612455010414,
 0.00497204065322876,
 -0.01935526356101036,
 0.007168988697230816,
 0.01749514788389206,
 -0.03901216760277748,
 0.004532148130238056,
 -0.007179043255746365,
 -0.038308341056108475,
 -0.0284145325422287,
 0.047216787934303284,
 -0.02974175103008747,
 0.021778441965579987,
 0.0026720319874584675,
 0.0508364737033844,
 -0.0026871138252317905,
 0.037343092262744904,
 -0.056547533720731735,
 0.022220848128199577,
 0.0012763162376359105,
 0.006781883537769318,
 -0.015132296830415726,
 -0.004069632850587368,
 0.0411236509680748,
 0.010456870310008526,
 0.028233548626303673,
 0.016278531402349472,
 -0.02384970709681511,
 -0.010044627822935581,
 -0.04942882061004639,
 0.006339477375149727,
 0.0269063301384449,
 -0.02519703470170498,
 -0.004235534928739071,
 -0.018349796533584595,
 0.026282940059900284,
 0.027972126379609108,
 0.0770188719034195,
 0.004416519310325384,
 -0.03320056200027466,
 -0.005540130194276571,
 -0.036679480224847794,
 0.015645086765289307,
 -0.017897335812449455,
 -0.01154277566820383,
 -0.0717502161860466,
 0.042792730033397675,
 0.0583171583712101,
 0.015755686908960342,
 0.03533215448260307,
 0.04842334985733032,
 0.021416474133729935,
 0.012648791074752808,
 -0.017756570130586624,
 0.005278708413243294,
 0.01151261106133461,
 -0.00702319573611021,
 0.016047272831201553,
 -0.002853016136214137,
 -0.025217143818736076,
 -0.010587580502033234,
 -0.0022070028353482485,
 0.012628681026399136,
 0.04118397831916809,
 0.03482941910624504,
 0.020209912210702896,
 0.007792378775775433,
 0.03993719816207886,
 0.05216369405388832,
 0.0006466417689807713,
 -0.04544716328382492,
 -0.007772269658744335,
 0.05719103291630745,
 0.06676309555768967,
 0.022522488608956337,
 -0.019003350287675858,
 0.03279837593436241,
 -0.0008489922620356083,
 0.0003899331495631486,
 -0.05441594123840332,
 -0.0366392619907856,
 0.02197953686118126,
 0.021878989413380623,
 0.0664815604686737,
 -0.001696727704256773,
 0.012367259711027145,
 0.0010142660466954112,
 -0.010808783583343029,
 0.02809278294444084,
 -0.025056269019842148,
 0.043154697865247726,
 0.027388954535126686,
 0.014116774313151836,
 -0.025056269019842148,
 0.058035627007484436,
 0.013433055952191353,
 -0.04194813594222069,
 -0.024915503337979317,
 -0.02660468965768814,
 -0.020833302289247513,
 0.0014704973436892033,
 -0.04096277803182602,
 0.047578759491443634,
 0.002460883464664221,
 -0.008315222337841988,
 0.04528629034757614,
 0.020974067971110344,
 -0.0095268115401268,
 -0.035030514001846313,
 -0.018500616773962975,
 0.01816881075501442,
 0.011130533181130886,
 -0.016117656603455544,
 0.016318749636411667,
 -0.028515079990029335,
 -0.04681460186839104,
 0.03939424455165863,
 0.06889468431472778,
 -0.05244522541761398,
 0.016761155799031258,
 -0.0448438823223114,
 -0.022160520777106285,
 0.03750396519899368,
 0.024392660707235336,
 0.04003774747252464,
 -0.004899144172668457,
 -0.07199152559041977,
 0.011854470707476139,
 0.03768495097756386,
 -0.00039590313099324703,
 0.023045333102345467,
 0.027610158547759056,
 0.00806888286024332,
 0.05312894284725189,
 0.021255599334836006,
 0.016560062766075134,
 0.06447062641382217,
 -0.010024518705904484,
 0.0020084227435290813,
 0.012678954750299454,
 0.009617303498089314,
 -0.030224375426769257,
 -0.012497970834374428,
 0.001496890909038484,
 0.01914411596953869,
 -0.00953686609864235,
 -0.018420178443193436,
 0.004909198731184006,
 0.01787722483277321,
 0.01499153207987547,
 -0.006254012696444988,
 -0.03452777862548828,
 -0.010029545985162258,
 -0.039715997874736786,
 -0.03806702792644501,
 -0.01050714310258627,
 -0.006344504654407501,
 0.01577579602599144,
 -0.03603598102927208,
 -0.009622330777347088,
 0.006691391114145517,
 -0.04878532141447067,
 -0.015504320152103901,
 -0.02340730093419552,
 0.05099735036492348,
 0.019093841314315796,
 -0.035533249378204346,
 -0.008335331454873085,
 -0.013171633705496788,
 0.0015270549338310957,
 -0.025619331747293472,
 -0.01024572178721428,
 -0.012970540672540665,
 -0.016499733552336693,
 -0.004449197091162205,
 0.005741223692893982,
 0.014488797634840012,
 -0.01181425154209137,
 0.006565707735717297,
 0.004851384554058313,
 0.01024572178721428,
 -0.009722878225147724,
 0.015343445353209972,
 0.034728873521089554,
 -0.006490297615528107,
 -0.03205432742834091,
 0.02123549021780491,
 0.019918326288461685,
 -0.02855529822409153,
 0.022964894771575928,
 -0.0047634057700634,
 -0.0074555473402142525,
 0.00017627114721108228,
 -0.0030189184471964836,
 0.017153289169073105,
 0.0016326290788128972,
 0.01136179082095623,
 0.00979326106607914,
 0.012136002071201801,
 -0.03915293514728546,
 -0.0328788124024868,
 0.027087314054369926,
 -0.004300890490412712,
 0.029781969264149666,
 0.04818204045295715,
 0.03977632522583008,
 0.00039904520963318646,
 0.03535226359963417,
 0.006047891452908516,
 0.043315570801496506,
 0.0038056974299252033,
 0.004884061869233847,
 0.025016050785779953,
 -0.023266535252332687,
 -0.026805784553289413,
 0.011854470707476139,
 0.0005247287335805595,
 -0.004290835931897163,
 -0.05711059644818306,
 0.056024692952632904,
 0.018963132053613663,
 4.114172043045983e-06,
 0.08268970996141434,
 0.007606367114931345,
 0.04206879064440727,
 -0.05300828814506531,
 0.01210583746433258,
 0.00870735477656126,
 -0.020129473879933357,
 -0.02974175103008747,
 0.012015345506370068,
 0.022240957245230675,
 0.028233548626303673,
 0.012719172984361649,
 -0.054818131029605865,
 -0.029017813503742218,
 0.03675992041826248,
 0.0679696574807167,
 0.022482270374894142,
 -0.022321395576000214,
 0.0024533425457775593,
 -0.015745632350444794,
 -0.0007823799969628453,
 -0.029922734946012497,
 -0.0020649803336709738,
 0.06382712721824646,
 0.011603103950619698,
 0.022381722927093506,
 -0.009360909461975098,
 0.008199593983590603,
 0.051238663494586945,
 0.0013611526228487492,
 0.035774558782577515,
 0.0530485063791275,
 0.0426318533718586,
 0.01697230339050293,
 -0.020310459658503532,
 0.062017280608415604,
 -0.008129211142659187,
 0.00026314990827813745,
 -0.01847045123577118,
 0.0063093132339417934,
 0.004672913812100887,
 0.03611642122268677,
 -0.012347150593996048,
 -0.03684035688638687,
 -0.04214923083782196,
 0.03042546845972538,
 -0.01815875619649887,
 -0.010185393504798412,
 -0.002910830546170473,
 0.00020376442989800125,
 0.027026986703276634,
 -0.0299026258289814,
 0.03193367272615433,
 -0.0269063301384449,
 -0.0010092387674376369,
 -0.008893366903066635,
 0.0037881017196923494,
 0.02767048589885235,
 0.049308162182569504,
 -0.045406945049762726,
 0.0119550172239542,
 -0.030606454238295555,
 0.024272004142403603,
 -0.01890280283987522,
 -0.009054241701960564,
 0.023487739264965057,
 -0.02011941932141781,
 -0.01008484698832035,
 0.03159181401133537,
 -0.030385250225663185,
 -0.011502556502819061,
 0.029239017516374588,
 0.00500471843406558,
 0.011331627145409584,
 -0.0017583125736564398,
 -0.036820247769355774,
 0.02049144357442856,
 -0.008752601221203804,
 -0.031209735199809074,
 -0.02756994031369686,
 -0.004826247692108154,
 -0.05345069244503975,
 0.005429528653621674,
 -0.035231608897447586,
 -0.03048579767346382,
 -0.01869165524840355,
 0.019375372678041458,
 0.07034255564212799,
 0.0004992778412997723,
 0.0010569985024631023,
 -0.022623036056756973,
 -0.00043769291369244456,
 -0.02616228349506855,
 -0.026745455339550972,
 0.016570117324590683,
 -0.017153289169073105,
 -0.009808342903852463,
 0.035392481833696365,
 -0.031068969517946243,
 0.000760385359171778,
 0.0014051418984308839,
 -0.03615663945674896,
 -0.009994354099035263,
 -0.009627358987927437,
 0.01423743087798357,
 0.04826247692108154,
 0.027248188853263855,
 0.024774737656116486,
 -0.04194813594222069,
 -0.013965954072773457,
 0.016027163714170456,
 0.0336429663002491,
 -0.0035518165677785873,
 0.0508364737033844,
 -0.008466042578220367,
 -0.005052478052675724,
 0.05522031709551811,
 -0.038227904587984085,
 0.02041100524365902,
 0.0016313722589984536,
 -0.0005687180091626942,
 -0.028193330392241478,
 -0.02662479877471924,
 -0.015142352320253849,
 0.022381722927093506,
 0.0314108282327652,
 -0.022783910855650902,
 -0.29842299222946167,
 -0.05546163022518158,
 -0.01748509332537651,
 -0.0156853049993515,
 0.036297403275966644,
 -0.003124492708593607,
 -0.016177983954548836,
 -0.017012523487210274,
 0.02415134757757187,
 0.008903421461582184,
 0.04987122491002083,
 -0.027690595015883446,
 0.022924676537513733,
 0.0004945647087879479,
 -0.04343622922897339,
 -0.02489539422094822,
 0.015896452590823174,
 -0.03191356360912323,
 -0.018701709806919098,
 -0.0381273552775383,
 0.04052037000656128,
 0.004016845487058163,
 -0.01853078044950962,
 0.014971422031521797,
 -0.0717904344201088,
 0.00016260307165794075,
 -0.016057327389717102,
 0.017997881397604942,
 0.007510847877711058,
 0.026343267410993576,
 0.009969217702746391,
 -0.005384282674640417,
 -0.045326508581638336,
 -0.02234150469303131,
 -0.028515079990029335,
 0.003602090058848262,
 0.017283998429775238,
 0.014378195628523827,
 -0.004607558250427246,
 -0.03086787462234497,
 -0.017002468928694725,
 0.04363732039928436,
 -0.0027801196556538343,
 -0.004124933388084173,
 0.026665018871426582,
 -0.014901039190590382,
 -0.02332686260342598,
 -0.003049082588404417,
 -0.0411035418510437,
 -0.06394778192043304,
 0.02863573655486107,
 -0.11261244118213654,
 0.022683363407850266,
 -0.025136705487966537,
 0.03191356360912323,
 0.023266535252332687,
 0.024573644623160362,
 -0.022442052140831947,
 -0.0426318533718586,
 -0.039575230330228806,
 -0.027791142463684082,
 0.009853589348495007,
 0.026122065261006355,
 -0.06611958891153336,
 0.007189097814261913,
 -0.026182392612099648,
 -0.0343669056892395,
 -0.0224219411611557,
 0.0011148129124194384,
 -0.016328804194927216,
 -0.029037922620773315,
 0.017374491319060326,
 0.014076555147767067,
 -0.029319453984498978,
 ...]

## 4. Criar função para buscar reviews relevantes a partir de uma query de **busca**

In [ ]:
import openai
import time

def text_embedding(text):
    """
    Recebe string, faz chamada ao modelo de embeddings da openai e retorna os embeddings da string
    """
    response = None
    retry_count = 0

    while response is None:
        try:
            response = openai.Embedding.create(engine="text-similarity-ada-001", input=text)
        except Exception as e:
            # Verifique se o erro é devido ao limite de chamadas
            if "call rate limit" in str(e):
                # Espere por um tempo exponencialmente crescente antes de tentar novamente
                time.sleep(2 ** retry_count)
                retry_count += 1
                continue
            else:
                # Se ocorrer outro tipo de erro, lide com ele de acordo com sua lógica de tratamento de erros
                print("Erro ao fazer chamada para a API: ", str(e))
                return None

    return response["data"][0]["embedding"]

In [ ]:
# Cria sample das linhas para fazer menos chamadas ao modelo
df_menor = df.sample(100);

In [ ]:
df_menor.shape

(100, 9)


(100, 8)



In [ ]:
# criar coluna com os embeddings
df_menor["embeddings"] = df_menor["texto_para_promtp"].apply(lambda linha: text_embedding(linha))



In [ ]:
from scipy import spatial

def strings_ranked_by_relatedness(
    query: str,
    df: pd.DataFrame,
    relatedness_fn=lambda x, y: 1 - spatial.distance.cosine(x, y),
    top_n: int = 10
) -> tuple[list[str], list[float]]:
    """
    Retorna as linhas do dataframe mais semelhantes a string de busca
    """

    query_embedding = text_embedding(query)

    strings_and_relatednesses = [ ]
    for i, row in df.iterrows():

        #relatedness = relatedness_fn(np.array(query_embedding).flatten(), np.array(row["embedding"]).flatten())
        relatedness = relatedness_fn(query_embedding, row["embeddings"])
        item = (row["texto_para_promtp"], relatedness)
        strings_and_relatednesses.append(item)

    strings_and_relatednesses.sort(key=lambda x: x[1], reverse=True)
    strings, relatednesses = zip(*strings_and_relatednesses)
    return strings[:top_n], relatednesses[:top_n]

In [ ]:
strings, relatednesses = strings_ranked_by_relatedness("Qual o game mais vendido",
                              df_menor)
for string, relatedness in zip(strings, relatednesses):
    print(f"{relatedness=:.3f}")
    display(string)

relatedness=0.811


'Quake II da publisher Activision vendeu 1.0 unidades na plataforma PC'

relatedness=0.809


'Doom da publisher id Software vendeu 2.0 unidades na plataforma PC'

relatedness=0.808


'Dark Souls II da publisher Bandai Namco Games vendeu 2.7 unidades na plataforma PC'

relatedness=0.808


'Black & White da publisher EA Games vendeu 2.0 unidades na plataforma PC'

relatedness=0.806


'Supreme Commander da publisher THQ vendeu 1.0 unidades na plataforma PC'

relatedness=0.806


'Spore da publisher Electronic Arts vendeu 2.0 unidades na plataforma PC'

relatedness=0.802


'Lords of the Realm II da publisher Impressions Game vendeu 2.5 unidades na plataforma PC'

relatedness=0.801


'Unreal da publisher GT Interactive vendeu 1.0 unidades na plataforma PC'

relatedness=0.801


'Guild Wars 2 da publisher NCsoft vendeu 5.0 unidades na plataforma PC'

relatedness=0.801


'Psychonauts da publisher THQ vendeu 1.0 unidades na plataforma PC'

**negrito**## 5. Criar prompt para perguntar ao gpt
* O modelo possui um limite de tokens, vamos verificar se o limite foi atingido ou não para criar o prompt

In [ ]:
!pip install tiktoken

In [ ]:
import tiktoken

def num_tokens(text: str) -> int:
    encoding = tiktoken.encoding_for_model("gpt-3.5-turbo-16k")
    return len(encoding.encode(text))

In [ ]:
num_tokens("oi brasil")

3

3

In [ ]:
num_tokens("Quais os jogos com mais sales de cada publisher")

11

In [ ]:
def montar_query(query: str, df: pd.DataFrame, token_budget: int) -> str:
    linhas, relatednesses = strings_ranked_by_relatedness(query, df)
    introducao = """
    Use o conteúdo abaixo com reviews de jogos para responder a pergunta a seguir.
    O conteúdo das vendas deve ser a sua única fonte para dar as respostas.
    Se você não conseguir encontrar a resposta no conteúdo, escreva "Não consegui encontrar uma resposta."
    """
    pergunta = f"\n\Pergunta: {query}"
    messagem = introducao
    for linha in linhas:
        proxima_linha = f'\n\nSeção com dados das sales de games:\n"""\n{linha}\n"""'
        if num_tokens(messagem + proxima_linha + pergunta) > token_budget:
            break
        else:
            messagem = messagem + proxima_linha

    return messagem + pergunta

In [ ]:
from pprint import pprint

prompt = montar_query(query="Qual game com mais sales de cada publisher",
             df=df_menor,
             token_budget=16000)
pprint(prompt)

('\n'
 '    Use o conteúdo abaixo com reviews de jogos para responder a pergunta a '
 'seguir.\n'
 '    O conteúdo das vendas deve ser a sua única fonte para dar as respostas.\n'
 '    Se você não conseguir encontrar a resposta no conteúdo, escreva "Não '
 'consegui encontrar uma resposta."\n'
 '    \n'
 '\n'
 'Seção com dados das sales de games:\n'
 '"""\n'
 'Doom da publisher id Software vendeu 2.0 unidades na plataforma PC\n'
 '"""\n'
 '\n'
 'Seção com dados das sales de games:\n'
 '"""\n'
 'Quake II da publisher Activision vendeu 1.0 unidades na plataforma PC\n'
 '"""\n'
 '\n'
 'Seção com dados das sales de games:\n'
 '"""\n'
 'Terraria da publisher Re-Logic vendeu 17.2 unidades na plataforma PC\n'
 '"""\n'
 '\n'
 'Seção com dados das sales de games:\n'
 '"""\n'
 'Unreal da publisher GT Interactive vendeu 1.0 unidades na plataforma PC\n'
 '"""\n'
 '\n'
 'Seção com dados das sales de games:\n'
 '"""\n'
 'Spore da publisher Electronic Arts vendeu 2.0 unidades na plataforma PC\n'
 '""

## 6. Enviar prompt para o gpt e mostrar as respostas

In [ ]:
def perguntar(
    query: str,
    df: pd.DataFrame = df,
    model: str = "gpt-35-turbo-16k",
    print_message: bool = False,
) -> str:
    message = montar_query(query, df, token_budget=16000)
    if print_message:
        print(message)
    messages = [
        {"role": "system", "content": "Você responde perguntas sobre sales de jogos."},
        {"role": "user", "content": message},
    ]
    response = openai.ChatCompletion.create(
        model=model,
        messages=messages,
        temperature=0,
        engine="gpt-35-turbo-16k"
    )
    response_message = response["choices"][0]["message"]["content"]
    return response_message

In [ ]:
resposta = perguntar("Qual game com mais sales de cada publisher",
    df_menor)
print(resposta)

- id Software: Doom (2.0 unidades vendidas)
- Activision: Quake II (1.0 unidades vendidas)
- Re-Logic: Terraria (17.2 unidades vendidas)
- GT Interactive: Unreal (1.0 unidades vendidas)
- Electronic Arts: Spore (2.0 unidades vendidas)
- EA Games: Black & White (2.0 unidades vendidas)
- THQ: Supreme Commander (1.0 unidades vendidas)
- Gathering of Developers: Tropico (1.0 unidades vendidas)
- Atari, Inc: The Witcher (2.0 unidades vendidas)
- Triternion: Mordhau (1.0 unidades vendidas)



2s
resposta = perguntar("Qual game com mais sales de cada publisher",
    df_menor)
print(resposta)
output
- id Software: Doom (2.0 unidades vendidas)
- Activision: Quake II (1.0 unidades vendidas)
- Re-Logic: Terraria (17.2 unidades vendidas)
- GT Interactive: Unreal (1.0 unidades vendidas)
- Electronic Arts: Spore (2.0 unidades vendidas)
- EA Games: Black & White (2.0 unidades vendidas)
- THQ: Supreme Commander (1.0 unidades vendidas)
- Gathering of Developers: Tropico (1.0 unidades vendidas)
- Atari, Inc: The Witcher (2.0 unidades vendidas)
- Triternion: Mordhau (1.0 unidades vendidas)

In [ ]:
montar_query(query="Qual game com mais sales de cada publisher",
             df=df_menor,
             token_budget=16000)

'\n    Use o conteúdo abaixo com reviews de jogos para responder a pergunta a seguir.\n    O conteúdo das vendas deve ser a sua única fonte para dar as respostas.\n    Se você não conseguir encontrar a resposta no conteúdo, escreva "Não consegui encontrar uma resposta."\n    \n\nSeção com dados das sales de games:\n"""\nDoom da publisher id Software vendeu 2.0 unidades na plataforma PC\n"""\n\nSeção com dados das sales de games:\n"""\nQuake II da publisher Activision vendeu 1.0 unidades na plataforma PC\n"""\n\nSeção com dados das sales de games:\n"""\nTerraria da publisher Re-Logic vendeu 17.2 unidades na plataforma PC\n"""\n\nSeção com dados das sales de games:\n"""\nUnreal da publisher GT Interactive vendeu 1.0 unidades na plataforma PC\n"""\n\nSeção com dados das sales de games:\n"""\nSpore da publisher Electronic Arts vendeu 2.0 unidades na plataforma PC\n"""\n\nSeção com dados das sales de games:\n"""\nBlack & White da publisher EA Games vendeu 2.0 unidades na plataforma PC\n"""\

\n    Use o conteúdo abaixo com reviews de jogos para responder a pergunta a seguir.\n    O conteúdo das vendas deve ser a sua única fonte para dar as respostas.\n    Se você não conseguir encontrar a resposta no conteúdo, escreva "Não consegui encontrar uma resposta."\n    \n\nSeção com dados das sales de games:\n"""\nDoom da publisher id Software vendeu 2.0 unidades na plataforma PC\n"""\n\nSeção com dados das sales de games:\n"""\nQuake II da publisher Activision vendeu 1.0 unidades na plataforma PC\n"""\n\nSeção com dados das sales de games:\n"""\nTerraria da publisher Re-Logic vendeu 17.2 unidades na plataforma PC\n"""\n\nSeção com dados das sales de games:\n"""\nUnreal da publisher GT Interactive vendeu 1.0 unidades na plataforma PC\n"""\n\nSeção com dados das sales de games:\n"""\nSpore da publisher Electronic Arts vendeu 2.0 unidades na plataforma PC\n"""\n\nSeção com dados das sales de games:\n"""\nBlack & White da publisher EA Games vendeu 2.0 unidades na plataforma PC\n"""\n

In [ ]:
resposta = perguntar("Existe algum jogo com 2 sales?",
    df_menor)
print(resposta)

Sim, vários jogos têm 2 sales. Alguns exemplos são Black & White da EA Games, Doom da id Software, Spore da Electronic Arts, Far Cry da Ubisoft, Baldur's Gate da Interplay Entertainment e Battlefield 1942 da Electronic Arts.


Sim, vários jogos têm 2 sales. Alguns exemplos são Black & White da EA Games, Doom da id Software, Spore da Electronic Arts, Far Cry da Ubisoft, Baldur's Gate da Interplay Entertainment e Battlefield 1942 da Electronic Arts.
